cpp/src/parquet/encoding.cc

```diff
@@ -1149,15 +1149,19 @@ template <>
 inline int DecodePlain<ByteArray>(const uint8_t* data, int64_t data_size, int num_values,
                                   int type_length, ByteArray* out) {
   int bytes_decoded = 0;
-  int increment;
   for (int i = 0; i < num_values; ++i) {
-    uint32_t len = out[i].len = arrow::util::SafeLoadAs<uint32_t>(data);
-    increment = static_cast<int>(sizeof(uint32_t) + len);
-    if (data_size < increment) ParquetException::EofException();
+    const uint32_t len = out[i].len = arrow::util::SafeLoadAs<uint32_t>(data);
+    const int64_t increment = static_cast<int64_t>(sizeof(uint32_t) + len);
+    if (ARROW_PREDICT_FALSE(data_size < increment)) {
+      ParquetException::EofException();
+    }
     out[i].ptr = data + sizeof(uint32_t);
     data += increment;
     data_size -= increment;
-    bytes_decoded += increment;
+    if (ARROW_PREDICT_FALSE(increment > INT_MAX - bytes_decoded)) {
+      throw ParquetException("BYTE_ARRAY chunk too large");
+    }
+    bytes_decoded += static_cast<int>(increment);
```

Comment:
Doesn't the `if` just above protect against it?

Reply:
Brain fart here, I'd move the check out of the loop, e.g. accumulate on int64_t and check only at the exit point.

Reply:
This also points to a design issue, since the input buffer size is typed with `int64_t` but it returns a `int`. This might be heavier to fix everywhere.